src/core/reader/types/values.rs

```diff
@@ -67,6 +67,27 @@ impl WasmReader<'_> {
         Ok(result)
     }
 
+    pub fn read_var_i33(&mut self) -> Result<i64> {
+        let mut result: i64 = 0;
+        let mut shift: u64 = 0;
+
+        let mut byte: i64;
+        loop {
+            byte = self.read_u8()? as i64;
+            result |= (byte & 0b0111_1111) << shift;
+            shift += 7;
+            if (byte & 0b1000_0000) == 0 {
+                break;
+            }
+        }
+
+        if shift < 33 && (byte & 0x40 != 0) {
+            result |= !0 << shift;
+        }
+
+        Ok(result)
+    }
```

Comment:
Are we confident this isn't the same as `read_var_i32` in practice? Since the first bit will always be 0 to be positive, do we need to store this in a i64?

Reply:
yeah they turned out to be different. We can't use i32 because the value inside might be larger than 2^31-1. We store it in an i64 in case the encoding turns out to be larger than u32 or negative, which we throw an error.

Reply:
Another reason they are using i encoding instead of u encoding is that unsigned single byte numbers might clash with the binary encoding of NumType etc. If we force that numbers larger than 2^7-1 are negative if its a single byte (first bit will be always 0 since that's how leb128 works) even if their unsigned repr can fit in a single byte, we enforce the blocktype to be at least two bytes.